### PR TITLE
[video_player] Fix some videos being rendered black on iOS

### PR DESCRIPTION
**Pull request by recastrodiaz**
I've found another instance of a video file that does not render using the video_player. The video settings are as follows:

```LOG
TX, TY, WIDTH, HEIGHT, ROTATION (degrees)
1080.0, 0.0, 848.0, 480.0, 90
```

Fixes https://github.com/flutter/flutter/issues/28831

For some reason the .tx value is incorrect. Setting it to 848.0 correctly renders the video.
I have refactored the code so .tx and .ty are always set (when rotationDegrees is not 0) regardless of their original values.

### CODE REVIEW

**Comment by jonasbark** [comment]
Applying that fix resulted a completely red output (not to confuse this with the Dart error view).

**Comment by recastrodiaz** [comment]
Interesting. @jonasbark Is the red screen happening to all of your videos? Otherwise, would you mind sharing one of them for me to test? You can send them to ml @ rodrigo . red if your prefer. Cheers
For context, I've been running this patch on several iOS devices with no reported issues so far.

**Comment by jonasbark** [comment]
@recastrodiaz sorry - nevermind. My testing branch didn't yet receive the fix where sendInitialized was called too early.

**Comment by recastrodiaz** [comment]
No problem. Glad it works!

**Comment by cyanglaz** [comment]
Thanks for the contribution!

Non-trivial fixes like this might be risky to land especially when we don't have an integration test for this plugin. And it will take more time to test things manually to make sure there is no regression. 

 I'm following the [initial PR review policy](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#the-review-process), this PR isn't trivial to review so I'm labeling it with "backlog" and we will prioritize according to the issue's priority.

related issue: https://github.com/flutter/flutter/issues/30967

**Comment by pmrajani** [comment]
@recastrodiaz i am using latest video_player: ^0.10.5+1 package.
But few videos are still see black screen in iOS (audio is working)
It's working for few videos and not working for some videos. all videos are in mp4 format.

Any solution for this?

**Comment by pmrajani** [comment]
@recastrodiaz still getting issue in iOS for black screen when playing video

**Comment by recastrodiaz** [comment]
@pmrajani can you test the following dependency in your pubspec.yaml and check if the videos are still black?

```YAML
dependencies:
  video_player:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player  
      ref: c0cede4e9e7aaf46e1747a4bdfdf2ad255aa080a
    version: ^0.10.5+1
```

**Comment by sowens-csd** [comment]
> ```yaml
>  video_player:
>     git:
>       url: https://github.com/recastrodiaz/plugins.git
>       path: packages/video_player  
>       ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163
>     version: ^0.10.5+2
> ```

I can recreate something that looks like the original issue. I'm testing on two different iPads. On one of them the video doesn't display, on the other it does. The one where it does not is older, an iPad 4th generation. The play button appears, sometimes a still image displays, sometimes the screen stays blank. Clicking play switches to the paused state and the sound starts playing but the video doesn't appear. There doesn't seem to be any error information in the log. 

I'm using the example code from the README of the video_player plugin. I had to choose a different video because sample-videos.com was slow or down today. Here's the video I used. https://docs.google.com/uc?export=download&id=1byHeESOMrMbmKKM78NLSl7pLe9YKsBpP 
I also tried 
https://file-examples.com/wp-content/uploads/2018/04/file_example_MOV_1280_1_4MB.mov 
but it has no sound so I can't tell if anything is happening, behaviour is otherwise the same. 

I had to change the pubspec reference a little. Here's what I used:
```
  video_player:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player/video_player
      ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163
    version: ^0.10.5+2
```

**Comment by recastrodiaz** [comment]
@sowends-csd is there a `TX, TY, WIDTH, HEIGHT, ROTATION (degrees)` message in your logs when using this branch?

**Comment by recastrodiaz** [comment]
Could you also test my latest develop2 branch on your older device?

Your pubspec should look something like this https://github.com/flutter/plugins/pull/1400#issuecomment-604587598
but you may need to override the video_player_platform_interface somehow (a user in that thread reported issues with resolving the dependencies)

**Comment by sowens-csd** [comment]
> @sowens-csd is there a `TX, TY, WIDTH, HEIGHT, ROTATION (degrees)` message in your logs when using this branch?

This is the log output when I play: 
2020-04-23 16:10:18.102480-0400 Runner[813:408664] VIDEO__ 0.000000, 0.000000, 720.000000, 1280.000000, 0

Looks like yes? 

I'll try the other issue you referenced.

**Comment by sowens-csd** [comment]
> Could you also test my latest develop2 branch on your older device?
> 
> Your pubspec should look something like this [#1400 (comment)](https://github.com/flutter/plugins/pull/1400#issuecomment-604587598)
> but you may need to override the video_player_platform_interface somehow (a user in that thread reported issues with resolving the dependencies)

I tried it but ran into the same version solving problem and don't have an easy way to resolve it. I think the issue is that in your develop2 branch the dependency to video_player_interface in the video_player/pubspec.yaml is to the release version instead of to the git version. If you change that I can test it. I could also fork the repo and do it but it would probably be better to keep the uncertainty to a minimum in these tests. 

Here's the line I think it's complaining about:
```
dependencies:
  meta: "^1.0.5"
  video_player_platform_interface: ^1.0.1
```

Probably should be:
```
  video_player_platform_interface:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player/video_player_platform_interface  
      ref: develop2
```
If you'd prefer I make the change in a fork just let me know. Alternately if you have another way to override the resolve problem I'm happy to try that.

**Comment by Maus3rSR** [comment]
> 
> 
> @pmrajani can you test the following dependency in your pubspec.yaml and check if the videos are still black?
> 
> ```yaml
> dependencies:
>   video_player:
>     git:
>       url: https://github.com/recastrodiaz/plugins.git
>       path: packages/video_player  
>       ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163
>     version: ^0.10.5+2
> ```

I have the same issue on iOS only. I'll test if I have still the issue with your fix. Thank you!

**Comment by recastrodiaz** [comment]
@Maus3rSR please check with the latest `develop2` branch too: https://github.com/recastrodiaz/plugins/tree/develop2 

Please bear in mind some people have had issues using this as it requires newer version of `video_player_platform_interface`

**Comment by Maus3rSR** [comment]
```
[enchere_equestre] flutter pub get
Running "flutter pub get" in enchere_equestre...                
Could not find a file named "packages/video_player/pubspec.yaml" in https://github.com/recastrodiaz/plugins.git 8344dcb20e2fe1884dba2745af1b0452df7fe163.
pub get failed (1; Could not find a file named "packages/video_player/pubspec.yaml" in https://github.com/recastrodiaz/plugins.git 8344dcb20e2fe1884dba2745af1b0452df7fe163.)
exit code 1
```

**Comment by Maus3rSR** [comment]
It's really a pain with this constraints version with `video_player_platform_interface`. Im facing the same issue than https://github.com/flutter/plugins/pull/1307#issuecomment-618648659

When I build I got an error than method `setSpeed` is undefined on `video_player_platform_interface`

I'm really stupefied why a bug like this on a video player plugin still exist.. it concern a lot of iOS devices...

**Comment by Maus3rSR** [comment]
Well...

I forked @recastrodiaz fork and remove all dependancy of video_player, video_player_web and video_player_platform_interface and setup my pubspec.yaml that I can build run and test it if it fix the issue....

```
  video_player_web :
    git:
      url: https://github.com/Maus3rSR/plugins.git
      path: packages/video_player/video_player_web   
      ref: develop2
  video_player_platform_interface:
    git:
      url: https://github.com/Maus3rSR/plugins.git
      path: packages/video_player/video_player_platform_interface  
      ref: develop2
  video_player:
    git:
      url: https://github.com/Maus3rSR/plugins.git
      path: packages/video_player/video_player  
      ref: develop2
```

And now on android, one it worked before testing this fork, got this exception:
```
I/ExoPlayerImpl(15367): Init eacaec5 [ExoPlayerLib/2.11.1] [dreamlte, SM-G950F, samsung, 28]
E/ExoPlayerImplInternal(15367): Internal runtime error.
E/ExoPlayerImplInternal(15367): java.lang.NullPointerException
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.util.Assertions.checkNotNull(Assertions.java:147)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.upstream.cache.CacheDataSink.<init>(CacheDataSink.java:106)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.upstream.cache.CacheDataSink.<init>(CacheDataSink.java:82)
E/ExoPlayerImplInternal(15367): 	at io.flutter.plugins.videoplayer.CacheDataSourceFactory.createDataSource(CacheDataSourceFactory.java:51)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.source.ProgressiveMediaSource.createPeriod(ProgressiveMediaSource.java:267)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.source.ExtractorMediaSource.createPeriod(ExtractorMediaSource.java:359)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.MediaPeriodHolder.createMediaPeriod(MediaPeriodHolder.java:409)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.MediaPeriodHolder.<init>(MediaPeriodHolder.java:97)
E/ExoPlayerImplInternal(15367): 	at com.google.android.exoplayer2.MediaPeriodQueue.enqueueNextMediaPeriodHolder(MediaPeriodQueue.java:155)
```

**Comment by recastrodiaz** [comment]
@sowens-csd thanks, I've updated my fork with your fix. The following should just work

``` YAML
dependencies:
  video_player:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player/video_player_platform_interface
      ref: develop2
```

@Maus3rSR not sure what your bug is about. I've updated ExoPlayer to 2.11.6 maybe give this a try again?

**Comment by liang530** [comment]
@recastrodiaz Hello, I use your version to test the video address. It's not a black screen, but a white screen. Is there a sound or a picture。

https://sp.huoyx.cn/video/6219/hsyx6219.mp4

**Comment by liang530** [comment]
Hello, I use your version to test the video address. It's not a black screen, but a white screen with only sound.

https://sp.huoyx.cn/video/6219/hsyx6219.mp4

**Comment by recastrodiaz** [comment]
@liang530 Are you able to play that mp4 with the iOS default player? I downloaded your video to a XR device and the video could not be played. Did you test with the `develop2` branch?

**Comment by diegomgarcia** [comment]
@recastrodiaz can you fix the path of your video_player_web in your pubspec develop2 branch ? as its pointing to the local path and breaking when we try to add the direct reference to your repository. 


`Error on line 36, column 11: Invalid description in the "video_player" pubspec on the "video_player_web" dependency: "../video_player_web" is a relative path, but this isn't a local pubspec.
   ╷
36 │     path: ../video_player_web`

**Comment by diegomgarcia** [comment]
@recastrodiaz just for feedback purpose I've tested with this version below and the blackscreen problem seems to be solved.

`video_player:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player/video_player
      ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163
    version: ^0.10.5+2`

**Comment by diegomgarcia** [comment]
@recastrodiaz just a little observation I've done more tests today and in one of the iPhones that I've tested the sound sometimes wasn't reproduced but I couldn't trace it if its something related with this version or something related with some configuration on this iPhone, with the oficial release 0.10.11+2 this issue with the sound doesn't seems to happen but the blackscreen is on it.

Any tips ?

**Comment by recastrodiaz** [comment]
@diegomgarcia `ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163` is a very old branch. You may want to try this instead: https://github.com/flutter/plugins/pull/1307#issuecomment-650454763

My `develop2` branch in line with the official release and has other features and bug fixes including the fix for the black screen. Let me know if this solves your no sound issue.

**Comment by diegomgarcia** [comment]
@recastrodiaz I was trying to use that one from develop2 but it has a physical reference to flutter web that is breaking the pub get. I've quoted you some messages before my test about that, I think you just have to update the link to point to the online repository instead of the physical one.

**Comment by recastrodiaz** [comment]
@diegomgarcia I've updated the reference for `video_player_web` please try again. https://github.com/recastrodiaz/plugins/commit/71cb6c04031f05578cf93bc06884261af7e98a87 Thanks!

**Comment by diegomgarcia** [comment]
@recastrodiaz thanks it works fine now and the volume has been fixed to.

This version should be promoted to the master release ;)

**Comment by recastrodiaz** [comment]
Is there any interest by the flutter plugin maintainers to approve this Pull Request? There are other fixes that need to be cherry-picked from `develop2` to properly fix "the black videos" issue on iOS, but given that this PR is more than 1y old, I'm hesitant to make any further changes to this branch and in general to opening any video_player plugin PRs.

For anyone else being hit by this issue, I recommend using for now:


``` YAML
dependencies:
  video_player:
    git:
      url: https://github.com/recastrodiaz/plugins.git
      path: packages/video_player/video_player
      ref: develop2
```

**Comment by yanivshaked** [comment]
Hi,

I tried using the following recommended settings:

> 
> ```yaml
> dependencies:
>   video_player:
>     git:
>       url: https://github.com/recastrodiaz/plugins.git
>       path: packages/video_player/video_player
>       ref: develop2
> ```

to solve the black rendered video on iOS, but I am getting the following error:
`
Could not load video tracks: 3 - Error Domain=AVFoundationErrorDomain Code=-11829 "Cannot Open" UserInfo={NSUnderlyingError=0x28046bc00 {Error Domain=NSOSStatusErrorDomain Code=-12848 "(null)"}, NSLocalizedFailureReason=This media may be damaged., NSURL=__VIMediaCache___:file:///private/var/mobile/Containers/Data/Application/<application_id>/tmp/61795987662__B2742873-45C2-49F7-9081-E1AABCEC67B5.MOV, NSLocalizedDescription=Cannot Open}

`

The same code is working perfectly on Android.
Appreciate any help!

**Comment by Colir** [comment]
@recastrodiaz Hi. i would like to test your package, but in my flutter app, i'm also using Chewie (https://pub.dev/packages/chewie). When validate my flutter pub get, i've this error
"Because chewie 0.9.10 depends on video_player >=0.7.0 <0.11.0 and no versions of chewie match >0.9.10 <0.10.0, chewie ^0.9.10 requires video_player from hosted.
So, because "myappname" depends on both video_player from git and chewie ^0.9.10, version solving failed.
"

Do you know if there are a trick to bypass this limitation ?

Thank you

**Comment by recastrodiaz** [comment]
Hi @Colir I'm not maintaining this PR at the moment. Have you tried using my latest "develop2" branch? https://github.com/flutter/plugins/pull/1307#issuecomment-650454763

**Comment by Colir** [comment]
@recastrodiaz in fact i'm a noob ! When i'm trying to use your branch, i've encounter an error 
`So, because "myappname" depends on both video_player from git and chewie ^0.9.10, version solving failed.`

**Comment by diegomgarcia** [comment]
@Colir I think you would have to clone the chewie repository and change its dependency to use the same one that @recastrodiaz recommended and use it int your project to. With this you should be good to go.

**Comment by danielwalczak** [comment]
Its working hotfix and its not super complicated, so why its not merged yet ?

**Review by Salakar — APPROVED** [review]
LGTM + confirmed working

File: packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m
Comment (jonasbark) [inline packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:183]: shouldn't this be removed or hidden behind a debug flag?

Comment (Salakar) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:183]: What would you suggest?

Normally I'd use `os_log` (iOS 10+) but given the iOS 9 minimum at the moment don't think we can here? Also there doesn't seem to be any standardised approach to native logging in any of the other plugins that I could see (a few NSLogs littered around) - perhaps this is something that could be addressed in separate PR? This log was here previously prior to this change albeit different wording so I didn't see any harm in it

Comment (jonasbark) [reply packages/video_player/video_player/ios/Classes/FLTVideoPlayerPlugin.m:183]: I agree - if there are NSLog statements already then a new PR makes more sense indeed. 

NSLog itself is fine IMO, it's just the fact that you'll get these messages during debugging (in Xcode only I guess) with no way to deactivate these.

**Comment by recastrodiaz** [comment]
Hi @Salakar, @jonasbark,  thanks for the review. This PR misses the case where `rotationDegrees == 180` so some videos might remain black. I fixed this in a different branch but never bothered to cherry-pick the change here. I'll push a commit shortly.

**Review by russellwheatley — APPROVED** [review]

**Comment by hnvn** [comment]
I am developing an app that lets users record their selfie video. The plugin cannot play this selfie video, black screen but there's still sound. Videos recorded by rear camera still work normally. I try to modify `fixTransform:` function, comment out all its codes and simply return `transform` variable, in this case the plugin can play both of videos recorded by front and rear camera. 

When I review the native codes using to record selfie videos, there's a different between using front and rear camera to record video:
```
//flip video output if front facing camera is selected
if self.currentCamera == .front {
   movieFileOutputConnection?.isVideoMirrored = true
}
```
The selfie video is marked as flipped vertically (to make mirror effect). I think codes of `fixTransform:` not cover this case and the black screen still happen.

**Comment by Abacaxi-Nelson** [comment]
+1 
Video on IOS are black, only sound..

**Comment by sahebsingh416** [comment]
> @recastrodiaz just for feedback purpose I've tested with this version below and the blackscreen problem seems to be solved.
> 
> ```
> video_player:
>     git:
>       url: https://github.com/recastrodiaz/plugins.git
>       path: packages/video_player/video_player
>       ref: 8344dcb20e2fe1884dba2745af1b0452df7fe163
>     version: ^0.10.5+2
> ```

What version of flick_video_player can we use with this?

**Comment by themadmrj** [comment]
Same here, some videos play sound but video is black...

**Comment by letsar** [comment]
This fix works really great but it needs to be rebased on the current version. @recastrodiaz, do you need help to do so, if yes, how can I help you?
In our project we had to fork the 2.1.6 version and use this fix, but it would be better if it could be merged.

**Comment by recastrodiaz** [comment]
@letsar I don't mind rebasing this PR as this is a simple fix, I was doing it for a long time, but there seems to be no interest in merging it from the Flutter team so I'll avoid doing any work until this PR will be approved. This PR is more than 2y old after all...

**Comment by letsar** [comment]
Thanks @recastrodiaz, I understand your point of view. @cyanglaz can we do something about this PR? It's an annoying bug with what it seems to be an easy fix. Is there any blocker from the Flutter team about this?
From our side, we have to fork this plugin and apply this fix, but it's not really maintainable. It would be really great if we could merge this in the codebase so that anyone could profit from it.

**Comment by stuartmorgan** [comment]
> Thanks @recastrodiaz, I understand your point of view. @cyanglaz can we do something about this PR?

This PR needs two things to move forward:
- Test coverage (as described in the contributing guide)
- Reviews (see below).

> It's an annoying bug with what it seems to be an easy fix. Is there any blocker from the Flutter team about this?

The issue from a review standpoint is that it's not at all clear from the comments in the PR that the fix is correct, and that it won't regress other videos, because (unless I'm missing something) it doesn't seem that the root cause here is understood. If I'm reading it correctly it's saying that there exist some videos whose transforms are wrong, therefore all transforms of all videos will be replaced.

How widespread is the problem? Do we know why it happens in the first place? What effect does this have on videos whose transforms are correct? How many videos has this been tested on to understand what the overall effect will be? Are there examples of other widely used video players that do this, demonstrating that the approach is sound? Currently, it doesn't seem that there are answers to any of these questions, which would mean that a reviewer would need to spend a substantial amount of time trying to find those answers. As such, this is in the backlog and subject to our [issue-based prioritization](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#the-review-process).

If someone can do the investigation above and update this PR with findings, that could certainly change the expected review effort such that it could be reviewed quickly.

**Comment by SahibG** [comment]
When will this be approved?

**Comment by SahibG** [comment]
@cyanglaz When will the bug fix be approved?

**Comment by stuartmorgan** [comment]
Please read the discussion above. The PR cannot be approved as-is.

**Comment by jasontcs** [comment]
Is the PR still not approved?

**Comment by stuartmorgan** [comment]
> > > What effect does this have on videos whose transforms are correct?
> > 
> > 
> > None. This is a follow up bugfix that missed 1 case (`180 degrees` else if clause) in a PR that has already been merged.
> 
> It's not just adding that case though, it's also making the entire block unconditional.

@recastrodiaz Can you address this question? It's still not at all clear to me why changing the behavior of the existing workaround to be applied in *far* more cases is necessary here.

**Comment by stuartmorgan** [comment]
> @recastrodiaz Can you address this question? It's still not at all clear to me why changing the behavior of the existing workaround to be applied in _far_ more cases is necessary here.

Absent a reply to this question, I don't see a way to move forward with this PR. Thank you for your contribution, and please don't hesitate to submit a new PR if you have the time to engage with a review. Thanks!

**Comment by stuartmorgan** [comment]
> I was unable to determine why sometimes some of the CGAffineTransform components are 0

I understand that. But I still don't understand why, if the goal of the PR is to fix an issue when the components are incorrectly zero, this PR *removes the existing check* that the workaround is only applied when the components are zero.

**Pull request closed without merging**
